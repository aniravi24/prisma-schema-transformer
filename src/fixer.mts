import p from '@prisma/internals';
const { getConfig, getDMMF } = p;
import { getConfigFile } from './config.mjs';
import {
  datasourceDeserializer,
  dmmfEnumsDeserializer,
  dmmfModelsDeserializer,
  generatorsDeserializer,
  Model,
} from './deserializer.mjs';
import { dmmfEnumTransformer, dmmfModelTransformer } from './transformer.mjs';

export async function fixPrismaFile(
  schema: string,
  deny: readonly string[] = [],
  configPath?: string,
) {
  const configFile = await getConfigFile(configPath);
  const denyList = [...new Set(deny.concat(configFile.deny))];

  const dmmf = await getDMMF({ datamodel: schema });
  const config = await getConfig({ datamodel: schema });

  const models = dmmf.datamodel.models as Model[];
  const datasources = config.datasources;
  const generators = config.generators;

  const filteredModels = models.filter((each) => !denyList.includes(each.name));
  const filteredEnums = dmmf.datamodel.enums.filter(
    (each) => !denyList.includes(each.name),
  );
  const transformedModels = dmmfModelTransformer(filteredModels, configFile);
  const transformedEnums = dmmfEnumTransformer(filteredEnums);

  return (
    await Promise.all([
      datasourceDeserializer(datasources),
      generatorsDeserializer(generators),
      dmmfModelsDeserializer(transformedModels),
      dmmfEnumsDeserializer(transformedEnums),
    ])
  )
    .filter((e) => e)
    .join('\n\n\n');
}
