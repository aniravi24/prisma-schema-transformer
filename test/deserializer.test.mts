import { readFileSync } from 'node:fs';
import { test, expect } from 'vitest';

import { fixPrismaFile } from '../src/fixer.mjs';
import { join } from 'node:path';

const getFile = (path: string = './fixtures/blog.prisma') => readFileSync(join(__dirname, path), 'utf-8');
const fix = (path: string) => fixPrismaFile(getFile(path)) ;
test('deserialized simple', () =>
  expect(fix('./fixtures/simple.prisma')).resolves.toMatchSnapshot());

test('deserialized schema', () =>
    expect(fix('./fixtures/schema.prisma')).resolves.toMatchSnapshot());


test('deserialized blog', () =>
    expect(fix('./fixtures/blog.prisma')).resolves.toMatchSnapshot());


test('bad config path', () =>
  expect(
    fixPrismaFile(getFile(),
      [],
        'not-exist.mjs'
    ),
  ).rejects.toMatchSnapshot());

test('bad config path extension', () =>
  expect(
    fixPrismaFile(getFile(),
      [],
        'not-exist.js'
    ),
  ).rejects.toMatchSnapshot());
