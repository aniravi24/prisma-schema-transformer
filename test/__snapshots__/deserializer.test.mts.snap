// Vitest Snapshot v1

exports[`bad config path 1`] = `[Error: bad config path "not-exist.mjs"]`;

exports[`bad config path extension 1`] = `[Error: config file extension should be ".mjs"]`;

exports[`deserialized blog 1`] = `
"
datasource postgresql {
	provider = \\"postgresql\\"
	url = env(\\"DATABASE_URL\\")
}


generator client {
  provider = \\"prisma-client-js\\"
}


model User {
    id String  @id   @default(dbgenerated(\\"gen_random_uuid()\\"))  
  email String @unique     
  name String?      
  posts Post[] @relation(\\"PostToUser\\")
  likes Like[] @relation(\\"LikeToUser\\")
}
model Post {
    id String  @id   @default(cuid())  
  createdAt DateTime     @default(now())  
  updatedAt DateTime      @updatedAt
  published Boolean      
  title String      
  content String?      
  authorId String?      
  author User? @relation(\\"PostToUser\\", fields: [authorId], references: [id])
  status PostStatus      
  likes Like[] @relation(\\"LikeToPost\\")
}
model Like {
    id String  @id   @default(cuid())  
  userId String      
  user User @relation(\\"LikeToUser\\", fields: [userId], references: [id])
  postId String      
  post Post @relation(\\"LikeToPost\\", fields: [postId], references: [id])
@@unique([userId, postId])
}



enum PostStatus {
	hidden
	visible
	@@map(\\"PostStatuses\\")
}"
`;

exports[`deserialized schema 1`] = `
"
datasource db {
	provider = \\"postgresql\\"
	url = env(\\"DATABASE_URL\\")
}


generator client {
  provider = \\"prisma-client-js\\"
  output = env(\\"FOO\\")
  another = \\"foo\\"
  anotherFoo = \\"foo\\"
}


model Employee {
    plantId String       @map(\\"PlantId\\")
  employeeId String @unique      @map(\\"EmployeeId\\")
  uni String       @map(\\"Uni\\")
  anothers Another[] @relation(\\"AnothersToEmployees\\")
  lastOne Last? @relation(\\"EmployeesToLast_PlantId_EmployeeOneId\\")
  lastTwo Last? @relation(\\"EmployeesToLast_PlantId_EmployeeTwoId\\")
@@map(\\"Employees\\")
@@unique([plantId, uni])
@@id([plantId, employeeId])
}
model Another {
  /// @Foo
/// @Bar
  anotherId String       @map(\\"AnotherId\\")
  plantId String       @map(\\"PlantId\\")
  employeeId String       @map(\\"EmployeeId\\")
  uni String       @map(\\"Uni\\")
  employee Employee @relation(\\"AnothersToEmployees\\", fields: [plantId,employeeId], references: [plantId,employeeId], onDelete: NoAction)
@@map(\\"Anothers\\")
@@id([plantId, anotherId])
}
model Last {
  /// @Foo
  plantId String       @map(\\"PlantId\\")
  id Int  @id    
  employeeOneId String       @map(\\"EmployeeOneId\\")
  employeeOne Employee @relation(\\"EmployeesToLast_PlantId_EmployeeOneId\\", fields: [plantId,employeeOneId], references: [plantId,employeeId], onDelete: NoAction)
  employeeTwoId String       @map(\\"EmployeeTwoId\\")
  employeeTwo Employee @relation(\\"EmployeesToLast_PlantId_EmployeeTwoId\\", fields: [plantId,employeeTwoId], references: [plantId,employeeId], onDelete: NoAction)
@@unique([plantId, employeeOneId])
@@unique([plantId, employeeTwoId])
}"
`;

exports[`deserialized simple 1`] = `
"model Post {
    id Int  @id   @default(autoincrement())  
  title String      
  content String?      
  published Boolean     @default(false)  
  authorId Int       @map(\\"author_id\\")
  author User @relation(\\"postsTousers\\", fields: [authorId], references: [id])
  createdAt DateTime     @default(now())   @map(\\"created_at\\")
  updatedAt DateTime     @default(now())  @updatedAt @map(\\"updated_at\\")
@@map(\\"posts\\")
}
model User {
    id Int  @id   @default(autoincrement())  
  email String @unique     
  weight Float      
  posts Post[] @relation(\\"postsTousers\\")
  createdAt DateTime     @default(now())   @map(\\"created_at\\")
  updatedAt DateTime     @default(now())  @updatedAt @map(\\"updated_at\\")
@@map(\\"users\\")
}"
`;
